src/pybind/mgr/dashboard/HACKING.rst

```diff
@@ -332,6 +332,87 @@ vs. ``it('image edit test' () => ...)``. As shown, the first example makes gramm
+Protractor runs tests against your application running in a real browser,
+interacting with it as a user would." `(src) <http://www.protractortest.org/#/>`__
+
+It requires a fully functional system and tests the interation of all components
+of the application (Ceph, back-end, front-end).
+E2E tests are designed to mimic the behavior of the user when interacting with the application
+- for example when it comes to workflows like creating/editing/deleting an item.
+Also the tests should verify that certain items are displayed as a user would see them
+when clicking through the UI.
+
+Angular Unit Tests:
+
+Unit tests, as the name suggests, are tests for smaller units of the code.
+Those tests are designed for testing all kinds of Angulars' components (e.g. services, pipes etc.).
+They do not require a connection to the backend and hence are independ of it,
+although the expected result of the backend can be verified using only the frontend.
+Data is either mocked or, in a simple case, contains a static input, a function call
+and an expected static output. More complex examples include the state of
+a component (attributes of the component class),that define how the output changes
+according to the given input.
+
+Which E2E/unit tests are considered to be valid?
+................................................
+
+This is not easy to answer, but new tests that are written in the same way as already existing
+dashboard tests should generally be considered valid.
+Unit tests should focus on the component to be tested.
+This is either an Angular component, directive, service, pipe, etc.
+
+E2E tests should focus on testing the functionality of the whole application.
+Approximately half of the E2E tests should verify the correctness of user visible elements.
+
+How should an E2E/unit test look like?
+......................................
+
+Unit tests should focus on the described purpose
+and shouldn't try to test other things in the same `it` block.
+
+E2E tests should contain a description that either verifies
+the correctness of a user visible element or a complete process
+like for example the creation/validation/deletion of a pool.
+
+What should an E2E/unit test cover?
+...................................
+
+E2E tests should mostly but not exclusively cover interaction with the backend.
+This way the interaction with the backend is utilized to write integration tests.
+
+A unit test should mostly cover critical or complex functionality
+of a component (Angular Components, Services, Pipes, Directives, etc).
+
+What should an E2E/unit test NOT cover?
+.......................................
+
+Avoid duplicate testing: do not write an E2E tests for what's already
+been covered as frontend-unit tests and vice versa.
+It may not be possible to completely avoid an overlap.
+
+Unit tests should not be used to extensively click through components and E2E tests
+shouldn't be used to extensively test a single component of Angular.
```

Comment:
> E2E tests should make use of the system scope and tell a story. They should not test frontend elements or a purely typescript related functionality of a form (as this would only be on the frontend scope).

This is a documentation about the status quo. We currently verify text that is displayed on the frontend that doesn't require the backend or Ceph to be shown. I'm not saying this is the best approach, but it's how we currently do things.

I also don't think this is the right section to show examples. The best practices/guideline would be better suited. But as this is not not about providing such examples, I'd also rather not add them do this documentation. The scope is to fill a void, not provide a complete set of examples we've ruled out before (also, we haven't talked about any best practices).

Reply:
The thing is that you can't say that the e2e test doesn't require ceph to be run, as the fronend would show errors if no connection would be there anymore - thus the test would fail. This information on how to really use the full power of an e2e testing will be provided in the future.

As per discussion I will provide information about the different tests scopes and usages for development in a separate issue afterwards, which will be discussed with a broader audience via a presentation.